stan/math/rev/fun/quad_form.hpp

```diff
@@ -93,6 +93,65 @@ class quad_form_vari : public vari {
 
   quad_form_vari_alloc<Ta, Ra, Ca, Tb, Rb, Cb>* impl_;
 };
+
+/**
+ * Return the quadratic form \f$ B^T A B \f$.
+ *
+ * Symmetry of the resulting matrix is not guaranteed due to numerical
+ * precision.
+ *
+ * @tparam T1 type of the first (square) matrix
+ * @tparam T2 type of the second matrix
+ *
+ * @param A square matrix
+ * @param B second matrix
+ * @return The quadratic form, which is a symmetric matrix.
+ * @throws std::invalid_argument if A is not square, or if A cannot be
+ * multiplied by B
+ */
+template <typename Mat1, typename Mat2,
+          require_all_matrix_t<Mat1, Mat2>* = nullptr,
+          require_any_var_matrix_t<Mat1, Mat2>* = nullptr>
+inline auto quad_form_impl(const Mat1& A, const Mat2& B, bool symmetric) {
+  check_square("quad_form", "A", A);
+  check_multiplicable("quad_form", "A", A, "B", B);
+
+  using return_t
+    = promote_var_matrix_t<decltype(value_of(B).transpose().eval() *
+				    value_of(A) *
+				    value_of(B).eval()), Mat1, Mat2>;
+
+  auto arena_A = to_arena(A);
+  auto arena_B = to_arena(B);
+
+  check_not_nan("multiply", "A", value_of(arena_A));
```

Comment:
Need to go through and make all the `check` functions work with `var<mat>` types. I'll make a list of the ones that need updated and add them to #2101. Here I just pass in the value.

Reply:
I think we should prefer passing values into check functions anyway. That way we don't need them to support `var`s. Also checks on value will be faster than ones on `Matrix<var>`.

Reply:
Maybe this is an expression thing, again. Like, if we pass in `MatrixXd`, then if we switch to value then we have to do two full reads of the matrix to do a check. If we pass in `Matrix<var, R, C>` it is one read. If we pass in an expression view of the values in a `Matrix<var, R, C>`, then it is still one read. This sound right?

Reply:
Yeah, but that is assuming the functions calling the check does not need to extract values anyway. Most functions do need values.

Reply:
> Most functions do need values.

That is true.

I guess if we just put a `auto x_val = value_of_rec(x)` inside all the checks then that would be good? If someone passes in values, that is good. If someone passes in `mat<var>`, it works via expressions.

Reply:
Yeah, it would work, but for `mat<var>` it would not be efficient. Also I am trying to say if we only pass in values we don't need to change all checks.

Reply:
Yeah, that makes sense. I don't want to accidentally make them slow.

I think our checks should work with `var<mat>` though. That means for ones that read values just adding a version that takes in a `var<mat>` and then calls itself with the value of the `var<mat>`.